sd_webui_bayesian_merger/tpe_optimiser.py

```diff
@@ -0,0 +1,63 @@
+from functools import partial
+from pathlib import Path
+
+from sd_webui_bayesian_merger.artist import draw_unet
+from sd_webui_bayesian_merger.merger import NUM_TOTAL_BLOCKS
+from sd_webui_bayesian_merger.optimiser import Optimiser, convergence_plot
+
+from hyperopt import Trials, hp, fmin, tpe, STATUS_OK
+
+
+class TPEOptimiser(Optimiser):
+
+    def _target_function(self, params):
+        res = self.sd_target_function(**params)
+        return {
+            'loss': 1./(res + 1e-10),
+            'status': STATUS_OK,
+            'params': params
+        }
+
+    def optimise(self) -> None:
+        # TODO: what if we want to optimise only certain blocks?
+        space = {f"block_{i}": hp.uniform(f'block_{i}', 0.0, 1.0) for i in range(NUM_TOTAL_BLOCKS)}
+        space["base_alpha"] = hp.uniform('base_alpha', 0.0, 1.0)
+
+        # this will do 20 warmup runs before optimising
+        self.trials = Trials()
+        tpe._default_n_startup_jobs = self.init_points
+        algo = partial(tpe.suggest, n_startup_jobs=self.init_points)
+        fmin(
+            self._target_function,
+            space=space,
+            algo=algo,
+            trials=self.trials,
+            max_evals=self.n_iters,
+        )
+
+        # clean up and remove the last merge
+        self._cleanup()
+
+    def _cleanup(self):
+        self.merger.remove_previous_ckpt(self.iteration + 1)
+
+    def postprocess(self) -> None:
+        scores = []
+        for i, res in enumerate(self.trials.losses()):
+            print(f"Iteration {i}: \n\t{-res}")
+            scores.append(-res)
+        best = self.trials.best_trial
+        print("Best:", best)
+
+        img_path = Path("logs", f"{self.merger.output_file.stem}.png")
+
+        convergence_plot(scores, figname=img_path)
+
+        unet_path = Path("logs", f"{self.merger.output_file.stem}-unet.png")
+        draw_unet(
+            best['params']['base_alpha'],
```

Comment:
this raises a `KeyError: 'params'`

perhaps that should be `best['spec']['base_alpha']`?

as for 
https://github.com/hyperopt/hyperopt/blob/6c5b33be93a71a4ae6a38f04a29b18020b78b702/hyperopt/base.py#L5-L12
```
- Trials - a list of documents including at least sub-documents:
    ['spec'] - the specification of hyper-parameters for a job
    ['result'] - the result of Domain.evaluate(). Typically includes:
        ['status'] - one of the STATUS_STRINGS
        ['loss'] - real-valued scalar that hyperopt is trying to minimize
    ['idxs'] - compressed representation of spec
    ['vals'] - compressed representation of spec
    ['tid'] - trial id (unique in Trials list)
```

Reply:
no, it's actually `best['result']['params']['base_alpha']`